File: Javascript/searching.js
let array = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];

// Linear Search
function linearSearch(arr, val) {
  for (let i = 0; i < arr.length; i++) {
    if (val == arr[i]) {
      return i;
    }
  }
  return -1;
}

// Binary Search
function binarySearch(arr, val) {
  let start = 0;
  let end = arr.length - 1;
  let middle = Math.round((start + end) / 2);
  for (let i = 0; i < Math.log(arr.length); i++) {

    if (start > end) {
      return -1;
    }
    if (val > arr[middle]) {
      start = middle;
      middle = Math.round((start + end) / 2);
    } else if (val < arr[middle]) {
      end = middle;
      middle = Math.round((start + end) / 2);
    } else {
      return middle;
    }
  }
}

File: C/searching.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int binarySearch(int arr[], int x, int length) {
  int l = 0;
  int r = length - 1;
  int middle = (l + r) / 2;

  for (int i = 0; i < length; i++) {
    if (l > r) { return -1; }
    if (x > arr[middle]) {
      l = middle;
      middle = floor(((double)l + (double)r) / 2.0);
    } else if (x < arr[middle]) {
      r = middle;
      middle = floor(((double)l + (double)r) / 2.0);
    } else {
      return middle;
    }
  }
  return 0;
}

int main(int argc, char** argv) {
  int x = 10;
  if (argc != 0) {
    x = atoi(argv[1]);
  }
  int arr[] = {1, 4, 10, 45, 46, 89, 101, 150};
  int length = sizeof(arr)/sizeof(int);

  int index = binarySearch(arr, x, length);
  printf("Binary Search for %d: %d\n", x, index);

  return 0;
}